### ISSUE TITLE
Getting a lot of timeout errors from redis

### ISSUE THREAD

**Issue by FrTomer** [StackExchange/StackExchange.Redis#1960]
Hi, 

I'm using StackExchangeRedis, Version=3.1.21.0 on my aspnet core app and I get many errors like the following :

`StackExchange.Redis.RedisTimeoutException: Timeout awaiting response (outbound=0KiB, inbound=0KiB, 8000ms elapsed, timeout is 8000ms), inst: 0, qu: 0, qs: 2, aw: False, rs: ReadAsync, ws: Idle, in: 38, in-pipe: 0, out-pipe: 0, serverEndpoint: Unspecified/host:6379, mgr: 10 of 10 available, clientName: 845df86cb4-l47l5, IOCP: (Busy=0,Free=1000,Min=1,Max=1000), WORKER: (Busy=78,Free=32689,Min=1,Max=32767), v: 2.0.593.37019 (Please take a look at this article for some common client-side issues that can cause timeouts: https://stackexchange.github.io/StackExchange.Redis/Timeouts)`

I use Async operation = _redisCache.GetAsync.

and I configure it on startup.cs as follows:

` services.AddStackExchangeRedisCache(option =>
            {
                if (redisSettings.Slave1 != null || redisSettings.Slave2 != null)
                {
                    option.ConfigurationOptions = new ConfigurationOptions
                    {
                        EndPoints =
                        {
                            {redisSettings.Host, redisSettings.Port},
                        },
                        SyncTimeout = 8000
                    };
                    if (!string.IsNullOrEmpty(redisSettings?.Slave1?.Host))
                        option.ConfigurationOptions.EndPoints.Add(redisSettings.Slave1.Host, redisSettings.Slave1.Port);
                    if (!string.IsNullOrEmpty(redisSettings?.Slave2?.Host))
                        option.ConfigurationOptions.EndPoints.Add(redisSettings.Slave2.Host, redisSettings.Slave2.Port);
                }
                else
                {
                    option.Configuration = $"{redisSettings.Host}:{redisSettings.Port}";
                }
            });`

Can someone please help me understand how to fix it ?

**Comment by NickCraver** [comment]
It looks like you're on `2.0.593.37019` of the client here (using some caching wrapper?) - in any case, I'd recommend updating the reference to our library to latest and seeing if the behavior shifts any. Overall, your CPU of thread pool is maxed out (see `WORKER: (Busy=78,Free=32689,Min=1,Max=32767)` in your exception). Either your CPU is idle with lots of hung threads, or it's pegged. Knowing which of these two is in play would help but looking at your queues and such, the bottleneck isn't in the client/connection, but in the overall processor & thread pool utilization.

**Comment by FrTomer** [comment]
Hi @NickCraver , thanks for your quick answer.
Can you please help me understand how do you see that the CPU is maxed out according to this : 
`(WORKER: (Busy=78,Free=32689,Min=1,Max=32767)`

In addition, I'm using asp net core 3.1 with Microsoft.Extensions.Caching.StackExchangeRedis 3.1.21 that uses StackExchange.Redis 2.0.593. Can I install the latest version of StackExchange.Redis on my project and Microsoft.Extensions.Caching.StackExchangeRedis will use it ?

So to sum up, "in the overall processor & thread pool utilization." means that the library requires an update or there's other things I should do on the client side?

**Comment by NickCraver** [comment]
> Can you please help me understand how do you see that the CPU is maxed out according to this

Yeah sure, what I mean is that's a lot of threads that are active. They're either busy doing work (pegging the CPU - you can check utilization and see if it's maxing out the CPU for example), or they're mostly waiting on something (e.g. sync over async operations). They're not _busy_, but they're _waiting_ in that case. Importantly though, your thread pool is limited and grows at a limited rate as well. If you suffer blocking like this in the app, you're waiting behind those threads to finish or the thread pool to grow - either can be painfully slower than needed to unblock. Note: .NET 6 has some enhancements for the blocking case that will more instantaneously grow the thread pool as an "out" here, but it's papering over the underlying issue.

> In addition, I'm using asp net core 3.1 with Microsoft.Extensions.Caching.StackExchangeRedis 3.1.21 that uses StackExchange.Redis 2.0.593. Can I install the latest version of StackExchange.Redis on my project and Microsoft.Extensions.Caching.StackExchangeRedis will use it ?

Yes, you can explicitly reference our latest and it'll get any benefits there without breaking. When we go to say 3.x, then there may be some incompatibilities but we're not even discussing that yet.

> So to sum up, "in the overall processor & thread pool utilization." means that the library requires an update or there's other things I should do on the client side?

There are very likely other things in the app causing the blocking, remember we're just _reporting_ the blocking with a timeout. It's likely a victim and think of it this way "would I have noticed a multi-second stall unless an error was thrown?" The answer to that for the vast majority of app developers is "...probably not", and so the timeout comes as a surprised but it's reporting a symptom that often isn't new.

Your best path if possible is to get a dump when this happens and pop it into any analysis tool - for example just drop the .dmp file into Visual Studio (literal drag/drop) and it'll give options to debug. You'd want to debug managed and look at the threads: see what they're all doing. It's either where they're busy (say encrypting or serialization data, something heavy), or where they're blocked (e.g. a lock pile-up, or a `.Wait`, or a `.Result` on an async or something of that variety).

**Comment by NickCraver** [comment]
Closing out to tidy up!
